Source/Core/VideoCommon/ShaderGenCommon.h

```diff
@@ -302,6 +303,15 @@ static const char s_shader_uniforms[] = "\tuint    components;\n"
                                         "\tuint4   xfmem_pack1[8];\n"
                                         "\tfloat4 " I_CACHED_TANGENT ";\n"
                                         "\tfloat4 " I_CACHED_BINORMAL ";\n"
+                                        "\tuint vertex_stride;\n"
+                                        "\tuint vertex_offset_rawnormal;\n"
+                                        "\tuint vertex_offset_rawtangent;\n"
+                                        "\tuint vertex_offset_rawbinormal;\n"
+                                        "\tuint vertex_offset_rawpos;\n"
+                                        "\tuint vertex_offset_posmtx;\n"
+                                        "\tuint vertex_offset_rawcolor0;\n"
+                                        "\tuint vertex_offset_rawcolor1;\n"
+                                        "\tuint4 vertex_offset_rawtex[2];\n"  // std140 is pain
```

Comment:
It seems a bit odd that these are included in the uniform for specialized shaders (and ubershaders that don't use the vertex loader), though I don't know if 64 additional bytes is really a problem.

Reply:
VS uniform size is currently 4032 bytes, and this makes it 4096 bytes, a ~1.6% increase.

If it becomes a big issue, our full vertex layout can be calculated from the component list plus ~17 bits of component size data (since the order is always the same), just a bit more messily.  We could put something like this at the top of `main` instead:
```glsl
// vs uniform: u32 component_sizes; // bits 0-16: rawtex, bit 17: position has 3 floats
uint vertex_stride = 0;
uint vertex_offset_posmtx;
uint vertex_offset_rawpos;
uint vertex_offset_rawnormal;
uint vertex_offset_rawtangent;
uint vertex_offset_rawbinormal;
uint vertex_offset_rawcolor0;
uint vertex_offset_rawcolor1;
uint vertex_offset_rawtex[8];
if (components & VB_HAS_POSMTX) {
  vertex_offset_posmtx = vertex_stride;
  vertex_stride++;
}
vertex_offset_position = vertex_stride;
vertex_stride += component_sizes & (1 << 17) ? 3 : 2;
if (components & VB_HAS_NORMAL) {
  vertex_offset_rawnormal = vertex_stride;
  vertex_stride += 3;
}
// ... binormal, tangent, color0, color1
for (int i = 0; i < 8; i++) {
  if (components & (VB_HAS_UV0 << i)) {
    vertex_offset_rawtex[i] = vertex_stride;
    vertex_stride += (component_sizes >> (2 * i)) & 3;
  }
}
```
Assuming the compiler recognizes those as uniform, AMD and Intel's ISAs probably wouldn't mind too much, it's just a bunch of sgprs on AMD (we usually have plenty to spare) and would take 2/128 gprs on Intel.  No clue how Nvidia and Apple would fare though, since I'm not sure those two have special storage for non-constant uniform values.

A second question would be how often do we upload new uniforms for something that doesn't affect non-uber shaders.  I would guess this happens more often in PS uniforms than VS uniforms though.  Might still be useful to have separate dirty flags for the two in the future.